### PR TITLE
ENH make sure packages in current repo data w/ features have versions without features

### PR DESCRIPTION
**Pull request by beckermr**
This PR makes sure that packages in the current repo data which have features also have version without features. This change makes sure that feature minimization works.

cc @isuruf @jjhelmus

### CODE REVIEW

**Comment by beckermr** [comment]
So the osx test that failed passes locally for me and the windows failures appear unrelated but I don't quite follow them. 

@marcelotrevisani do you know what is going on here?

**Comment by beckermr** [comment]
Ah it appears all of these tests are failing on master too. So I think this one is ready to go!

File: conda_build/index.py [hunk 1]
```diff
@@ -715,14 +715,22 @@ def _shard_newest_packages(subdir, r, pins=None):
...
                 version = r.find_matches(MatchSpec('%s=%s' % (g_name, pin_value)))[0].version
                 matches.update(r.find_matches(MatchSpec('%s=%s' % (g_name, version))))
-        groups[g_name] = matches
+        groups[g_name] = set(
+            m
+            for m in matches
+            if not (m.track_features or m.features)
+        )
+        for m in matches:
+            print(m, m.track_features, m.features, m in groups[g_name])
```
Comment (beckermr) [inline conda_build/index.py:724]:
```suggestion
```

File: conda_build/index.py [hunk 2]
```diff
@@ -715,14 +715,22 @@ def _shard_newest_packages(subdir, r, pins=None):
             for pin_value in pins[g_name]:
                 version = r.find_matches(MatchSpec('%s=%s' % (g_name, pin_value)))[0].version
                 matches.update(r.find_matches(MatchSpec('%s=%s' % (g_name, version))))
-        groups[g_name] = matches
+        groups[g_name] = set(
+            m
+            for m in matches
+            if not (m.track_features or m.features)
+        )
+        for m in matches:
```
Comment (beckermr) [inline conda_build/index.py:723]:
```suggestion
```

File: conda_build/index.py [hunk 3]
```diff
@@ -715,14 +715,22 @@ def _shard_newest_packages(subdir, r, pins=None):
...
+            print(m, m.track_features, m.features, m in groups[g_name])
+
     new_r = _get_resolve_object(subdir, precs=[pkg for group in groups.values() for pkg in group])
     return set(_add_missing_deps(new_r, r))
 
 
 def _build_current_repodata(subdir, repodata, pins):
     r = _get_resolve_object(subdir, repodata=repodata)
     keep_pkgs = _shard_newest_packages(subdir, r, pins)
+    print([k.name for k in keep_pkgs])
```
Comment (beckermr) [inline conda_build/index.py:733]:
```suggestion
```

**Comment by beckermr** [comment]
Bump here! I fixed one bug in the pr and it appears to be working now in my tests of the new sysroot stuff for conda forge.

**Comment by jjhelmus** [comment]
I'm not certain excluding packages with features entirely from `current_repodata.json` is a good idea.  Excluding packages from `current_repodata` could result in packages without features having un-met dependencies because the dependency has a feature.

I think a better solution is to treat features similar to "pins". That is, if package with features is included in `current_repodata` the newest version package without features should also be include, provided that one exists.

**Comment by beckermr** [comment]
> I'm not certain excluding packages with features entirely from current_repodata.json is a good idea. Excluding packages from current_repodata could result in packages without features having un-met dependencies because the dependency has a feature.

So, in this case, the solver will fall back to the full repodata, but that is a performance hit. Is the performance the concern here?

**Comment by beckermr** [comment]
> I think a better solution is to treat features similar to "pins". That is, if package with features is included in current_repodata the newest version package without features should also be include, provided that one exists.

This should work.

**Comment by jjhelmus** [comment]
> So, in this case, the solver will fall back to the full repodata, but that is a performance hit. Is the performance the concern here?

I'm not worried about performance, more my worry is that if all of the packages for a particular name have a feature then that entire package is excluded `current_repodata` which may cause the solver to prefer the solution from `current_repodata` over something that is more logical.  

As an example:  Currently `conda create -n test blas=*=openblas numpy` will give a solution using `current_repodata`.  With this proposed change the `blas-1.0-openblas` package would not be included in `current_repodata` and conda should fall back to the full repodata.  

My concern is that there will be some other solution that conda does find in current_repodata for these edge cases that will make no sense.  That said, I cannot provide a non-contrived concrete example.  Expanding, `current_repodata` to always include a non-feature package seems like a lower risk than removing all feature packages.

**Comment by beckermr** [comment]
Ahhhhh thank you! That makes sense to me!

**Comment by beckermr** [comment]
ok @jjhelmus this is ready for another look!

File: conda_build/index.py [hunk 4]
```diff
@@ -696,6 +696,33 @@ def _add_missing_deps(new_r, original_r):
...
 
+def _add_prev_ver_for_features(new_r, orig_r):
+    expanded_groups = copy.deepcopy(new_r.groups)
+    for g_name in new_r.groups:
+        if not any(m.track_features or m.features for m in new_r.groups[g_name]):
+            # no features so skip
+            continue
+
+        # versions are sorted here so this is the latest
+        latest_version = VersionOrder("%s" % new_r.groups[g_name][0].version)
```
Comment (beckermr) [inline conda_build/index.py:707]:
```suggestion
        latest_version = VersionOrder(new_r.groups[g_name][0].version)
```

File: conda_build/index.py [hunk 5]
```diff
@@ -696,6 +696,33 @@ def _add_missing_deps(new_r, original_r):
...
+        if g_name in orig_r.groups:
+            # now we iterate through the list to find the next to latest
+            # without a feature
+            keep_m = None
+            for i in range(len(orig_r.groups[g_name])):
+                _m = orig_r.groups[g_name][i]
+                if (
+                    VersionOrder(_m.version) <= latest_version and
```
Comment (beckermr) [inline conda_build/index.py:715]:
```suggestion
                    VersionOrder(str(_m.version)) <= latest_version and
```

**Comment by beckermr** [comment]
Bump here @jjhelmus

**Comment by beckermr** [comment]
Bump here

**Comment by jjhelmus** [comment]
LGTM thanks again for putting this together @beckermr

**Pull request merged**